ondevice_learning.qmd

```diff
@@ -1,61 +1,630 @@
+
+NWP will anticipate the next word the user is trying to type based on the previous one. SC gives inline suggestions to speed up the typing based on each character. OTF will re-rank the proposed next words based on the active typing process. All three of these models need to run quickly on the edge and federated learning can accelerate training on the users data. However, uploading every word a user typed to the cloud for training would be a massive privacy violation. Therefore, federated learning with an emphasis on differential privacy protects the user while still enabling a better user experience. 
+
+![Examples of Google G Board Features](images/ondevice_gboard_example.png)
+
+To accomplish this goal, Google employed their algorithm DP-FTRL which provides a formal guarantee that trained models will not memorize specific user data or identities. DP-FTRL combined with secure aggregation, a strategy of encrypting model updates, provides an optimal balance of privacy and utility. Furthermore, adaptive clipping is applied in the aggregation process to limit the impact of individual users on the global model. Through a combination of all of these techniques, Google can continuously refine their keyboard while preserving user privacy in a formally provable way.
+
+![Google's System Design for Differential Privacy in G Board](images/ondevice_gboard_approach.png)
+
+### Benchmarking for Federated Learning: MedPerf
+
+One of the richest examples of data on the edge is medical devices. These devices store some of the most personal data on users but offer huge advances in personalized treatment and better accuracy in medical AI. Given these two factors, medical devices are the perfect use case for federated learning. [MedPerf](https://doi.org/10.1038/s42256-023-00652-2) is an open source platform used to benchmark models using federated evaluation [@karargyris2023federated]. Instead of just training models via federated learning, MedPerf takes the model to edge devices to test it against personalized data while preserving privacy. In this way a benchmark committee can evaluate various models in the real world on edge devices while still preserving patient anonymity. 
 
 ## Security Concerns
 
-Explanation: Security is a significant concern for any system that performs learning on-device, as it may expose vulnerabilities.
+Performing machine learning model training and adaptation on end-user devices also introduces security risks that must be addressed. Some key security concerns include:
+
+- **Exposure of private data**: Training data may be leaked or stolen from devices
```

Comment:
Might be worth it to expand on model inversion attacks which could potentially leak things like iCloud photos taken from iPhones
https://arxiv.org/abs/2304.01669

Reply:
Oh nice, thanks for the suggestion and link! 👍